wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,2592 @@
+			if ( null !== $pre ) {
+				return $pre;
+			}
+		}
+		$this->pdo_fetch_mode = $fetch_mode;
+		$this->mysql_query    = $query;
+
+		try {
+			// Parse the MySQL query.
+			$lexer  = new WP_MySQL_Lexer( $query );
+			$tokens = $lexer->remaining_tokens();
+
+			$parser = new WP_MySQL_Parser( self::$grammar, $tokens );
+			$ast    = $parser->parse();
+
+			if ( null === $ast ) {
+				throw new Exception( 'Failed to parse the MySQL query.' );
+			}
+
+			// Handle transaction commands.
+			$child = $ast->get_child();
+			if ( $child instanceof WP_Parser_Node && 'beginWork' === $child->rule_name ) {
+				return $this->begin_transaction();
+			}
+
+			if ( $child instanceof WP_Parser_Node && 'simpleStatement' === $child->rule_name ) {
+				$subchild = $child->get_child_node( 'transactionOrLockingStatement' );
+				if ( null !== $subchild ) {
+					$tokens = $subchild->get_descendant_tokens();
+					$token1 = $tokens[0];
+					$token2 = $tokens[1] ?? null;
+					if (
+						WP_MySQL_Lexer::START_SYMBOL === $token1->id
+						&& WP_MySQL_Lexer::TRANSACTION_SYMBOL === $token2->id
+					) {
+						return $this->begin_transaction();
+					}
+
+					if (
+						WP_MySQL_Lexer::BEGIN_SYMBOL === $token1->id
+					) {
+						return $this->begin_transaction();
+					}
+
+					if (
+						WP_MySQL_Lexer::COMMIT_SYMBOL === $token1->id
+					) {
+						return $this->commit();
+					}
+
+					if (
+						WP_MySQL_Lexer::ROLLBACK_SYMBOL === $token1->id
+					) {
+						return $this->rollback();
+					}
+				}
+			}
+
+			// Perform all the queries in a nested transaction.
+			$this->begin_transaction();
+
+			do {
+				$error = null;
+				try {
+					$this->execute_mysql_query( $ast );
+				} catch ( PDOException $error ) {
+					if ( $error->getCode() !== self::SQLITE_BUSY ) {
+						throw $error;
+					}
+				}
+			} while ( $error );
+
+			if ( function_exists( 'do_action' ) ) {
+				/**
+				 * Notifies that a query has been translated and executed.
+				 *
+				 * @param string $query The executed SQL query.
+				 * @param string $query_type The type of the SQL query (e.g. SELECT, INSERT, UPDATE, DELETE).
+				 * @param string $table_name The name of the table affected by the SQL query.
+				 * @param array $insert_columns The columns affected by the INSERT query (if applicable).
+				 * @param int $last_insert_id The ID of the last inserted row (if applicable).
+				 * @param int $affected_rows The number of affected rows (if applicable).
+				 *
+				 * @since 0.1.0
+				 */
+				do_action(
+					'sqlite_translated_query_executed',
+					$this->mysql_query,
+					$this->query_type,
+					$this->table_name,
+					$this->insert_columns,
+					$this->last_insert_id,
+					$this->affected_rows
+				);
+			}
+
+			// Commit the nested transaction.
+			$this->commit();
+
+			return $this->return_value;
+		} catch ( Exception $err ) {
+			// Rollback the nested transaction.
+			$this->rollback();
```

Comment:
would this throw if an exception is thrown before the first `begin_transaction()`  call?

Reply:
Good catch, adding a check.